spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala

```diff
@@ -0,0 +1,201 @@
+  require(outputFormats.length == outputs.length, s"IRGraph: outputFormats" +
+    s"length ${inputFormats.length} should be same with input nodes length ${outputs.length}")
+
+  val allNodes = new ArrayBuffer[Node[IRElement[T]]]
+  private var graph: Graph[T] = null
+
+  init()
+  private def init() : Unit = {
+    getNodes(inputs, allNodes)
+    // reminder: some output nodes may not be searched from inputs
+    outputs.foreach(node => {
+      if (!allNodes.contains(node)) allNodes.append(node)
+    })
+  }
+
+  private def getNodes(inputs: Seq[Node[IRElement[T]]],
+                       nodesBuffer: ArrayBuffer[Node[IRElement[T]]]): Unit = {
+    if (inputs.length == 0) return
+    inputs.foreach(node => {
+      if (!nodesBuffer.contains(node)) {
+        nodesBuffer.append(node)
+        getNodes(node.nextNodes, nodesBuffer)
+      }
+    })
+  }
+
+  override def updateOutput(input: Activity): Activity = {
+    if (graph == null) {
+      throw new UnsupportedOperationException("forward not supported, Please build graph first")
+    }
+    initFwdPrimitives(input)
+    output = graph.updateOutput(input)
+    output
+  }
+
+  override def updateGradInput(input: Activity, gradOutput: Activity): Activity = {
+    if (graph == null) {
+      throw new UnsupportedOperationException("backward not supported, Please build graph first")
+    }
+    initBwdPrimitives()
+    gradInput = graph.updateGradInput(input, gradOutput)
+    gradInput
+  }
+
+  override def accGradParameters(input: Activity, gradOutput: Activity): Unit = {
+    if (graph == null) {
+      throw new UnsupportedOperationException("backward not supported, Please build graph first")
+    }
+    initGradWPrimitives()
+    graph.accGradParameters(input, gradOutput)
+  }
+
+  def build(dnnMode: Boolean = false): Unit = {
+    graph = new IRConverter[T](this).toGraph(dnnMode)
+  }
+
+  override def parameters(): (Array[Tensor[T]], Array[Tensor[T]]) = {
+    graph.parameters()
+  }
+
+  override def training(): this.type = {
+    train = true
+    graph.training()
+    this
+  }
+
+  /**
+    * Set the module to evaluate mode
+    * @return
+    */
+  override def evaluate(): this.type = {
+    train = false
+    graph.evaluate()
+    this
+  }
+
+  private def initFwdPrimitives(input: Activity): Unit = {
+    if (!initFwd && graph.isInstanceOf[DnnGraph]) {
+      val inputMemory = new Array[MemoryData](inputFormats.length)
+      if (input.isInstanceOf[Tensor[T]]) {
+        inputMemory(0) = HeapData(input.toTensor[T].size(), inputFormats(0))
+      } else {
+        val tensors = input.toTable
+        require(tensors.length() == inputFormats.length, s"table input length " +
+          s"${tensors.length()} should be same with inputFormats length ${inputFormats.length}")
+        tensors.foreach(t => {
+          require(t._2.isInstanceOf[Tensor[T]],
+            "Only support input just contains tensor, but here get Table in input")
```

Comment:
Only support input with tensor type, table not supported

Reply:
ok, changed